### PR TITLE
living entity allow attribute registration

### PR DESCRIPTION
**Pull request by yannicklamprecht**

### CODE REVIEW

File: Spigot-Server-Patches/0596-living-entity-allow-attribute-registration.patch [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+-    private final Map<AttributeBase, AttributeModifiable> b = Maps.newHashMap();
++    private final Map<AttributeBase, AttributeModifiable> b = Maps.newHashMap(); private final Map<AttributeBase, AttributeModifiable> attributeMap = b; // Paper - OBFHELPER
+     private final Set<AttributeModifiable> c = Sets.newHashSet();
+     private final AttributeProvider d;
+ 
+@@ -130,4 +130,10 @@ public class AttributeMapBase {
+         }
+ 
+     }
++    // Paper - start
```
Comment (Proximyst) [inline Spigot-Server-Patches/0596-living-entity-allow-attribute-registration.patch]: A little extra whitespace before here.

File: Spigot-Server-Patches/0596-living-entity-allow-attribute-registration.patch [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
++    public void registerAttribute(AttributeBase attributeBase){
```
Comment (Proximyst) [inline Spigot-Server-Patches/0596-living-entity-allow-attribute-registration.patch]:
```suggestion
+    public void registerAttribute(AttributeBase attributeBase) {
```

**Review by Proximyst — DISMISSED** [review]
Very minor. LGTM!

File: Spigot-Server-Patches/0596-living-entity-allow-attribute-registration.patch (hunk 1 above)
Comment (yannicklamprecht) [reply Spigot-Server-Patches/0596-living-entity-allow-attribute-registration.patch]: Added before and also after.

**Review by Proximyst — APPROVED** [review]

**Comment by electronicboy** [comment]
How does this work with entities being un/reloaded?

**Comment by yannicklamprecht** [comment]
> How does this work with entities being un/reloaded?

From my knowledge they are persistent.

AttributeMapBase#c() return the attribute nbt tag. 
```
nbttagcompound.set("Attributes", this.getAttributeMap().c());
```
This is called in LivingEntity#saveData()
And loaded into attributeMap on entity load.

**Comment by yannicklamprecht** [comment]
Any complains here?

**Comment by MiniDigger** [comment]
why was this closed?

**Comment by yannicklamprecht** [comment]
rebased.

**Comment by yannicklamprecht** [comment]
@electronicboy You're fine with that? Especially with how the the un-/load of entities is affected.

> > How does this work with entities being un/reloaded?
> 
> From my knowledge they are persistent.
> 
> AttributeMapBase#c() return the attribute nbt tag.
> 
> ```
> nbttagcompound.set("Attributes", this.getAttributeMap().c());
> ```
> 
> This is called in LivingEntity#saveData()
> And loaded into attributeMap on entity load.

**Review by electronicboy — APPROVED** [review]

**Comment by electronicboy** [comment]
My heads currently spinning so not too sure I wanna hit the button to actually merge, but, from what I can see, they'll load back just fine, but, there is the requirement that you ensure that your attributes are actually registered onLoad or whatever, not entirely too fond of the idea of attribute data being lost if a plugin fails to load, however, but am not too sure if this is even avoidable without just ignoring unknown attributes and ensuring that we just stick them somewhere we can forcefully throw them back in, which would prevent some cases of upset on stuff like migration from forge at the cost of keeping invalid attribute data in memory, which, meh

**Comment by yannicklamprecht** [comment]
> My heads currently spinning so not too sure I wanna hit the button to actually merge, but, from what I can see, they'll load back just fine, but, there is the requirement that you ensure that your attributes are actually registered onLoad or whatever, not entirely too fond of the idea of attribute data being lost if a plugin fails to load, however, but am not too sure if this is even avoidable without just ignoring unknown attributes and ensuring that we just stick them somewhere we can forcefully throw them back in, which would prevent some cases of upset on stuff like migration from forge at the cost of keeping invalid attribute data in memory, which, meh

Currently only vanilla attributes can be "registered" here. If we want to support real custom attributes, like modded stuff, then we should address that in some other pr.

The term of "register" is more of "Add a vanilla attribute to an entity that doesn't have it by default when attributes were initialized."

Take your time or let others decide.

**Comment by electronicboy** [comment]
ah, I see, I misread ome intent

**Comment by yannicklamprecht** [comment]
Thank you very much.

**Comment by yannicklamprecht** [comment]
Ah one thing we forgot, was rebasing. Should be enough to rename these two patches as they apply successfully. Just to keep the right order. 😅

**Comment by electronicboy** [comment]
The bot should auto do that but I think it's stuck behind a pending upstream

**Pull request merged**